File: src/main/java/me/shedaniel/linkie/commands/HelpCommand.java
package me.shedaniel.linkie.commands;

import me.shedaniel.linkie.CommandBase;
import me.shedaniel.linkie.InvalidUsageException;
import me.shedaniel.linkie.LinkieBot;
import org.javacord.api.entity.message.MessageAuthor;
import org.javacord.api.entity.message.embed.EmbedBuilder;
import org.javacord.api.event.message.MessageCreateEvent;

import java.util.concurrent.ScheduledExecutorService;

public class HelpCommand implements CommandBase {
    @Override
    public void execute(ScheduledExecutorService service, MessageCreateEvent event, MessageAuthor author, String cmd, String[] args) {
        if (args.length != 0)
            throw new InvalidUsageException("+" + cmd);
        String prefix = LinkieBot.getInstance().commandApi.getPrefix(!event.isServerMessage() || (event.getServer().isPresent() && event.getServer().get().getId() != 432055962233470986l));
        event.getChannel().sendMessage(new EmbedBuilder()
                .setTitle("Linkie Help Command")
                .setFooter("Requested by " + author.getDiscriminatedName(), author.getAvatar())
                .addField("Help Command", "Value: " + prefix + "help, " + prefix + "?, " + prefix + "commands\nDisplays this message.")
                .addField("Fabric Api Versions Command", "Value: " + prefix + "fabricapi\nCheck Fabric API versions for every mc version.")
                .addField("Yarn Class Command", "Value: " + prefix + "yc\nCheck yarn 1.2.5 mappings.")
                .addField("Yarn Field Command", "Value: " + prefix + "yf\nCheck yarn 1.2.5 mappings.")
                .addField("Yarn Method Command", "Value: " + prefix + "ym\nCheck yarn 1.2.5 mappings.")
                .addField("Yarn Version Command", "Value: " + prefix + "yv\nCheck yarn 1.2.5 mappings last fetch time.")
                .setTimestampToNow()
        );
    }
}

File: src/main/kotlin/me/shedaniel/linkie/commands/HelpCommand.kt
package me.shedaniel.linkie.commands

import discord4j.core.`object`.entity.MessageChannel
import discord4j.core.`object`.entity.User
import discord4j.core.event.domain.message.MessageCreateEvent
import me.shedaniel.linkie.*
import java.util.*

object HelpCommand : CommandBase {
    override fun execute(event: MessageCreateEvent, user: User, cmd: String, args: Array<String>, channel: MessageChannel) {
        if (args.isNotEmpty())
            throw InvalidUsageException("!$cmd")
        val prefix = commandApi.getPrefix(event.guildId.orElse(null)?.asLong() == 432055962233470986L)
        val commandCategories = CommandCategory.getValues(event.guildId.orElse(null)).filter { c ->
            commandApi.commands.filter { it.key.getCategory() == c && it.key.getName() != null && it.key.getDescription() != null && it.value.isNotEmpty() }.isNotEmpty()
        }
        commandCategories.forEachIndexed { index, category ->
            channel.createEmbed {
                if (index == 0) it.setTitle("Linkie Help Command")
                else category.description?.let(it::setTitle)
                if (index == commandCategories.size - 1) {
                    it.setFooter("Requested by " + user.discriminatedName, user.avatarUrl)
                    it.setTimestampToNow()
                }
                commandApi.commands.filter { it.key.getCategory() == category && it.key.getName() != null && it.key.getDescription() != null && it.value.isNotEmpty() }
                        .toSortedMap(Comparator.comparing<CommandBase, String> { it.getName()!! })
                        .forEach { cmd, values ->
                            if (values.isEmpty()) return@forEach
                            val name = cmd.getName() ?: return@forEach
                            val desc = cmd.getDescription() ?: return@forEach
                            it.addInlineField(name, "Value: ${values.joinToString { "$prefix$it" }}\n$desc")
                        }
            }.subscribe()
        }
    }

    override fun getDescription(): String? = "Displays this message."
    override fun getName(): String? = "Help Command"
}